### PR TITLE
I12 genericize exclude models

### PR DESCRIPTION
**Pull request by kirkkwang**
ref #12 
- add configuration to `#exclude_models`
- update specs
- update config template

### CODE REVIEW

File: lib/generators/iiif_print/templates/config/initializers/iiif_print.rb [hunk 1]
```diff
@@ -1,6 +1,17 @@
...
+  #
   # Add WorkTypes into the Array to use PageDerivativeService
   # @example
   #   # config.work_types_for_derivative_service = [GenericWork, Image]
   # config.work_types_for_derivative_service = []
+
+  # Add models to be excluded from search so the user
+  # would not see them in the search results
+  # @example
+  #   # config.model_name_solr_field_values = ['Generic Work']
```
Comment (ShanaLMoore) [inline lib/generators/iiif_print/templates/config/initializers/iiif_print.rb:12]: Does this mean that generic work would excluded?

File: lib/generators/iiif_print/templates/config/initializers/iiif_print.rb [hunk 2]
```diff
@@ -1,6 +1,16 @@
...
   # Add WorkTypes into the Array to use PageDerivativeService
   # @example
   #   # config.work_types_for_derivative_service = [GenericWork, Image]
+  #
   # config.work_types_for_derivative_service = []
+
+  # Add models to be excluded from search so the user
+  # would not see them in the search results.
+  # @example
+  #   # config.excluded_model_names = [Generic Work, Image]
```
Comment (jeremyf) [inline lib/generators/iiif_print/templates/config/initializers/iiif_print.rb:13]:
```suggestion
  #   # config.excluded_model_names = ["Generic Work", "Image"]
```

File: spec/search_builders/custom_search_builder_spec.rb [hunk 3]
```diff
@@ -25,10 +25,20 @@
...
 
-    before { subject.exclude_models(solr_parameters) }
-    it 'adds the facet fields to solr_parameters' do
-      expect(solr_parameters[:fq]).to be_truthy
-      expect(solr_parameters[:fq]).to include('-human_readable_type_sim:"Newspaper Title"')
+    context 'with configured model name' do
+      before do
+        class ExcludedModel; end
+        class AnotherExcludedModel; end
+        IiifPrint.config.excluded_model_names = [ExcludedModel, AnotherExcludedModel]
```
Comment (jeremyf) [inline spec/search_builders/custom_search_builder_spec.rb:32]:
```suggestion
        IiifPrint.config.excluded_model_names = ["Excluded Model Name", "Another Excluded Model"]
```

File: spec/search_builders/custom_search_builder_spec.rb [hunk 4]
```diff
@@ -25,10 +25,20 @@
...
+        subject.exclude_models(solr_parameters)
+      end
+
+      it 'adds the facet fields to solr_parameters with configured key' do
+        expect(solr_parameters[:fq]).to be_truthy
+        expect(solr_parameters[:fq]).to(
+          include("-has_model_ssim:\"ExcludedModel\"", "-has_model_ssim:\"AnotherExcludedModel\"")
```
Comment (jeremyf) [inline spec/search_builders/custom_search_builder_spec.rb:39]:
```suggestion
          include("-has_model_ssim:\"Excluded Model Name\"", "-has_model_ssim:\"Another Excluded Model\"")
```

**Review by jeremyf — CHANGES_REQUESTED** [review]
I made some suggestions to change the names to reflect that they are strings and not classes.

**Review by ShanaLMoore — APPROVED** [review]

File: app/search_builders/concerns/iiif_print/exclude_models.rb [hunk 5]
```diff
@@ -6,10 +6,11 @@ module ExcludeModels
 
     def exclude_models(solr_parameters)
```
Comment (jeremyf) [inline app/search_builders/concerns/iiif_print/exclude_models.rb:7]:
```suggestion
    def exclude_models(solr_parameters, config: IiifPrint.config)
```

File: spec/search_builders/custom_search_builder_spec.rb [hunk 6]
```diff
@@ -25,10 +25,33 @@
...
     end
 
-    before { subject.exclude_models(solr_parameters) }
-    it 'adds the facet fields to solr_parameters' do
-      expect(solr_parameters[:fq]).to be_truthy
-      expect(solr_parameters[:fq]).to include('-human_readable_type_sim:"Newspaper Title"')
+    context 'with configured model name solr field values' do
+      before do
+        IiifPrint.config.excluded_model_name_solr_field_values = ['Excluded Model', 'Another Excluded Model']
+        subject.exclude_models(solr_parameters)
```
Comment (jeremyf) [inline spec/search_builders/custom_search_builder_spec.rb:31]:
```suggestion
        config = IiifPrint::Configuration.new.tap { |c| c.excluded_model_name_solr_field_values = ['Excluded Model', 'Another Excluded Model'] }
        subject.exclude_models(solr_parameters, config: config)
```

**Review by jeremyf — CHANGES_REQUESTED** [review]
By setting the config in the before block, if we run specs in random order, sometimes they will fail because we're making assumptions about the default configuration.  Instead favor either:

1. Dependency injection of the configuration for this method (as partially suggested)
2. Favor an around spec block that captures the initial global config, changes the global config, runs the test, and then restores the global config.

I strongly advise going with option 1.

File: spec/search_builders/custom_search_builder_spec.rb [hunk 7]
```diff
@@ -25,10 +25,36 @@
...
+        )
+      end
+
+      context 'with configured model name solr field key' do
+        before do
+          config = IiifPrint::Configuration.new.tap do |c|
+            c.excluded_model_name_solr_field_values = ['ExcludedModel', 'AnotherExcludedModel']
+            c.excluded_model_name_solr_field_key = 'has_model_ssim'
+          end
+          byebug
```
Comment (jeremyf) [inline spec/search_builders/custom_search_builder_spec.rb:47]: Please remove this

**Review by jeremyf — APPROVED** [review]

**Pull request merged**